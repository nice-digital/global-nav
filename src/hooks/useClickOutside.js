import { useEffect, useRef } from "react";

function useClickOutside(action) {
	const ref = useRef(null);

	// Returns true if it is a DOM element for IE11
	// see https://stackoverflow.com/questions/384286/how-do-you-check-if-a-javascript-object-is-a-dom-object
	function isElement(o) {
		return typeof HTMLElement === "object"
			? o instanceof HTMLElement //DOM2
			: o &&
					typeof o === "object" &&
					o !== null &&
					o.nodeType === 1 &&
					typeof o.nodeName === "string";
	}

	const handleClickOutside = (event) => {
		// detects that the clicked element is not inside the referenced element
		if (
			ref.current &&
			isElement(ref.current) &&
			!ref.current.contains(event.target)
		) {
			action(true);
		}
	};

	useEffect(() => {
		document.addEventListener("click", handleClickOutside, true);
		return () => {
			document.removeEventListener("click", handleClickOutside, true);
		};
	}, [handleClickOutside]);

	return {
		ref,
	};
}

export default useClickOutside;
