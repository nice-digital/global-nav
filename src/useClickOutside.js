import { useRef, useEffect } from "react";

function useClickOutside() {
	const ref = useRef(null);

	const handleClickOutside = (event) => {
		if (ref.current && !ref.current.contains(event.target)) {
			console.log("outside click ");
			// setidOfOpenDropdown(null);
		} else {
			console.log("inside click ");
		}
	};

	useEffect(() => {
		document.addEventListener("click", handleClickOutside, true);
		return () => {
			document.removeEventListener("click", handleClickOutside, true);
		};
	}, []);

	return {
		ref,
	};
}

export default useClickOutside;
