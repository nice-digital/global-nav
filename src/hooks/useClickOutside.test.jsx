import React from "react";
import useClickOutside from "./useClickOutside";
import { shallow, mount } from "enzyme";
import toJson from "enzyme-to-json";

const testFunction = jest.fn();

function App() {
	const testVariable = "test";
	const { ref } = useClickOutside(testVariable, testFunction);

	// function wasClickedOutside(result) {
	// 	if (result == true) context.setidOfOpenDropdown(null);
	// }
	return (
		<>
			<div ref={ref}>
				<button>inside</button>
			</div>
			<button>outside</button>
		</>
	);
}

describe("useClickOutside hook", () => {
	const defaultProps = {};

	beforeEach(() => {
		// window.dataLayer = [];
	});

	afterAll(() => {
		// Cleanup
		// delete window.dataLayer;
	});

	it("Renders without crashing", () => {
		const wrapper = shallow(<App {...defaultProps} />);
		// console.log(wrapper.debug());
		expect(wrapper).toHaveLength(1);
	});

	it("Matches snapshot", () => {
		const wrapper = shallow(<App {...defaultProps} />);
		expect(toJson(wrapper)).toMatchSnapshot();
	});

	it("Returns true when button outside of referenced element is clicked", () => {
		const wrapper = mount(<App {...defaultProps} />);
		const insideButton = wrapper.findWhere((node) => {
			return (
				node.type() === "button" && node.name() && node.text() === "inside"
			);
		});
		const outsideButton = wrapper.findWhere((node) => {
			return (
				node.type() === "button" && node.name() && node.text() === "outside"
			);
		});

		insideButton.simulate("click");
		console.log("###", insideButton.debug());
		console.log("###", outsideButton.debug());

		// expect(testFunction).toHaveBeenCalled();

		expect(true).toBe(false);
	});
});
