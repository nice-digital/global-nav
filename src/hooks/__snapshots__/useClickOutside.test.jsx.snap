// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useClickOutside hook Matches snapshot 1`] = `
<Fragment>
  <div>
    <button>
      inside
    </button>
  </div>
  <button>
    outside
  </button>
</Fragment>
`;
